src/NovelRT/Experimental/Graphics/Vulkan/VulkanGraphicsBuffer.cpp

```diff
@@ -0,0 +1,237 @@
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        void* pDestination = nullptr;
+
+        VkResult mapMemoryResult =
+            vkMapMemory(vulkanDevice, vulkanDeviceMemory, GetOffset(), GetSize(), 0, &pDestination);
+
+        if (mapMemoryResult != VK_SUCCESS)
+        {
+            // TODO: Make a real exception.
+            throw std::runtime_error("Failed to map Vulkan memory to the CPU! Reason: " +
+                                     std::to_string(mapMemoryResult));
+        }
+
+        return reinterpret_cast<uint8_t*>(pDestination) + rangeOffset;
+    }
+
+    const void* VulkanGraphicsBuffer::MapForReadUntyped()
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        void* pDestination = nullptr;
+
+        VkResult mapMemoryResult =
+            vkMapMemory(vulkanDevice, vulkanDeviceMemory, GetOffset(), GetSize(), 0, &pDestination);
+
+        if (mapMemoryResult != VK_SUCCESS)
+        {
+            // TODO: Make a real exception.
+            throw std::runtime_error("Failed to map Vulkan memory to the CPU! Reason: " +
+                                     std::to_string(mapMemoryResult));
+        }
+
+        uint64_t nonCoherentAtomSize =
+            device->GetAdapter()->GetVulkanPhysicalDeviceProperties().limits.nonCoherentAtomSize;
+
+        size_t offset = GetOffset();
+        size_t size = (GetSize() + nonCoherentAtomSize - 1) & ~(nonCoherentAtomSize - 1);
+
+        VkMappedMemoryRange mappedMemoryRange{};
+        mappedMemoryRange.sType = VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE;
+        mappedMemoryRange.memory = vulkanDeviceMemory;
+        mappedMemoryRange.offset = offset;
+        mappedMemoryRange.size = size;
+
+        VkResult invalidateMappedMemoryRangesResult =
+            vkInvalidateMappedMemoryRanges(device->GetVulkanDevice(), 1, &mappedMemoryRange);
+
+        if (invalidateMappedMemoryRangesResult != VK_SUCCESS)
+        {
+            // TODO: Make a real exception.
+            throw std::runtime_error("Failed to map the Vulkan memory for read only!");
+        }
+
+        return pDestination;
+    }
+
+    const void* VulkanGraphicsBuffer::MapForReadUntyped(size_t readRangeOffset, size_t readRangeLength)
+    {
+        std::shared_ptr<VulkanGraphicsDevice> device = GetDevice();
+
+        VkDevice vulkanDevice = device->GetVulkanDevice();
+        VkDeviceMemory vulkanDeviceMemory = GetBlock()->GetVulkanDeviceMemory();
+
+        void* pDestination = nullptr;
+
+        VkResult mapMemoryResult =
+            vkMapMemory(vulkanDevice, vulkanDeviceMemory, GetOffset(), GetSize(), 0, &pDestination);
+
+        if (mapMemoryResult != VK_SUCCESS)
+        {
+            // TODO: Make a real exception.
+            throw std::runtime_error("Failed to map Vulkan memory to the CPU!");
```

Comment:
Boy, Khronos is gonna get mad if you only keep giving them "fake" exceptions 😝

Reply:
See previous todos. :D